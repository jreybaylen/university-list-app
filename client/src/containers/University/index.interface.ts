import { MouseEvent } from 'react'

import { UniversityProps } from '@interface/api.interface'

interface ModifiedUniversityProps extends UniversityProps {
    onSelect?(param: MouseEvent<HTMLDivElement>): void
}

export type {
    ModifiedUniversityProps
}