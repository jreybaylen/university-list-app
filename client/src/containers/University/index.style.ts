import { StypesProps } from '@interface/style.interface'

const styles: StypesProps = {
    university: {
        width: '500px',
        height: '500px'
    },
    country: {
        fontSize: '.85rem',
        margin: '0 0 8px 10px'
    },
    name: {
        fontWeight: 500,
        color: '#3e3e3e',
        margin: '0 0 15px 10px'
    },
    websiteContainer: {
        display: 'flex'
    }
}

export {
    styles
}