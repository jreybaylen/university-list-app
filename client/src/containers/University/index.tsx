import { useEffect, useState } from 'react'

import { styles } from './index.style'
import { UniversityProps } from '@interface/api.interface'

import { Card } from '@components/Card'

function University (props: UniversityProps): JSX.Element {
    const [ name, setName ] = useState<string>('')
    const [ country, setCountry ] = useState<string>('')
    const [ websites, setWebsites ] = useState<Array<string>>([])

    useEffect(() => {
        const { name, country, ...rest } = props

        setName(name)
        setCountry(country)
        setWebsites(rest.web_pages)
    }, [ props ])
    
    const universityElement = (
        <Card>
            <p style={ styles.country }>{ country }</p>
            <h3 style={ styles.name }>{ name }</h3>
            <div style={ styles.websiteContainer }>
                { websites.map(
                    (website: string) => (
                        <a 
                            key={ website }
                            target="_blank"
                            rel="noreferrer"
                            href={ website }
                            style={ styles.website }
                        >
                            { website }
                        </a>
                    )
                ) }
            </div>
        </Card>
    )

    return universityElement
}

export default University