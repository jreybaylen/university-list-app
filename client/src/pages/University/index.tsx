import axios from 'axios'
import { useParams, useHistory } from 'react-router-dom'
import { lazy, useEffect, useState, useCallback, Fragment } from 'react'

import { styles } from './index.style'
import { UniversityProps, APIResponseProps } from '@interface/api.interface'

import { Information } from '@components/Information'

const WebsiteLink = lazy(() => import('@container/WebsiteLink'))

function University (): JSX.Element {
    const history = useHistory()
    const { name } = useParams<{ name: string }>()
    const [ university, setUniversity ] = useState<UniversityProps>()
    const handleGetSpecificUniversity = useCallback(async () => {
        try {
            const { data }: APIResponseProps = await axios.get(`/search?name=${ name }`)
            const [ universityInfo ] = data

            setUniversity(universityInfo)
        } catch (error: any) {
            console.error('Error Found: ', error)
        }
    }, [ name ])
    const handleOpenWebsite = (website: string) => {
        window.open(website, '_blank')
    }
    const handleBack = () => {
        history.replace('/')
    }

    useEffect(() => {
        handleGetSpecificUniversity()
    }, [handleGetSpecificUniversity ])

    const universityElement = (
        <Fragment>
            <button style={ styles.back } onClick={ handleBack }>
                Back
            </button>
            <div style={ styles.container }>
                <div style={ styles.content }>
                    <h1 style={ styles.heading }>{ university?.name }</h1>
                </div>
            </div>
            <div style={ styles.information }>
                <Information
                    title="University"
                    content={ `${ university?.name }, ${ university?.alpha_two_code }` || '' }
                />
                <Information
                    title="Country"
                    content={ university?.country || '' }
                />
                <Information title="Website(s)">
                    { university?.web_pages.map(
                        (website: string) => (
                            <WebsiteLink
                                key={ website }
                                website={ website }
                                onSelect={ handleOpenWebsite }
                            />
                        )
                    ) }
                </Information>
                <Information title="Domain(s)">
                    { university?.domains.map(
                        (domain: string) => (
                            <p key={ domain }>{ domain }</p>
                        )
                    ) }
                </Information>
            </div>
        </Fragment>
    )

    return universityElement
}

export default University