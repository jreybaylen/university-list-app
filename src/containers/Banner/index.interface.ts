interface BannerProps {
    title: string
}

export type {
    BannerProps
}