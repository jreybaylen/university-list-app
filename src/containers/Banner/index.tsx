import { Fragment } from 'react'
import { useHistory } from 'react-router-dom'

import { styles } from './index.style'
import { BannerProps } from './index.interface'

function Banner (props: BannerProps): JSX.Element {
    const history = useHistory()
    const handleBack = () => {
        history.replace('/')
    }

    const bannerElement = (
        <Fragment>
            <button style={ styles.back } onClick={ handleBack }>
                Back
            </button>
            <div style={ styles.container }>
                <h1 style={ styles.heading }>{ props.title }</h1>
            </div>
        </Fragment>
    )

    return bannerElement
}

export default Banner