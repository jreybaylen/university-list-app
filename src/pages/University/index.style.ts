import univBGSrc from '@assets/univ-bg.jpg'
import { StypesProps } from '@interface/style.interface'

const styles: StypesProps = {
    back: {
        top: 20,
        left: 20,
        color: '#fff',
        cursor: 'pointer',
        borderRadius: '25px',
        padding: '10px 20px',
        position: 'absolute',
        border: '1px solid #fff',
        background: 'rgba(0, 0, 0, 0.5)'
    },
    container: {
        marginBottom: 50,
        padding: '150px 0',
        backgroundSize: 'cover',
        backgroundPosition: 'center',
        backgroundRepeat: 'no-repeat',
        boxShadow: '0px 0px 15px 5px #ccc',
        backgroundImage: `url(${ univBGSrc })`
    },
    banner: {
        margin: '10% 0'
    },
    heading: {
        width: '80%',
        color: '#fff',
        margin: 'auto',
        fontWeight: 500,
        borderRadius: 5,
        maxWidth: '1000px',
        textAlign: 'center',
        padding: '30px 15px',
        border: '2px solid #fff',
        background: 'rgba(0, 0, 0, 0.6)',
        boxShadow: 'rgb(0 0 0 / 20%) 0px 4px 8px 0px'
    },
    information: {
        width: '90%',
        margin: 'auto',
        maxWidth: '1000px'
    }
}

export {
    styles
}