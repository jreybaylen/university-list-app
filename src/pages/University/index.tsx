import axios from 'axios'
import { useParams } from 'react-router-dom'
import { lazy, useEffect, useState, useCallback, Fragment } from 'react'

import { styles } from './index.style'
import { StorageKeyProps } from '@util/index.interface'
import { ProfileProps } from '@interface/profile.interface'
import { UniversityProps, APIResponseProps } from '@interface/api.interface'

import { Information, Button, WebsiteLink } from '@components/index'

const Banner = lazy(() => import('@container/Banner'))

function University (): JSX.Element {
    const { name } = useParams<{ name: string }>()
    const [ buttonLabel, setButtonLabel ] = useState<string>('Add')
    const [ universityKey, setUniversityKey ] = useState<string>('')
    const [ university, setUniversity ] = useState<UniversityProps>()
    const handleGetListofUniv = useCallback(async (init?: boolean, univKeyToStore?: string) => {
        const { environment } = await import('@config/environment')
        const { saveUnivStorage } = environment as { saveUnivStorage: StorageKeyProps }
        const { setDataToStorage, getUniversityListByUser, getAuthUserFromStorage } = await import('@util/index')
        const { username } = getAuthUserFromStorage() as ProfileProps
        const [ saveUniversityByUser, univListFromStorageByUser ] = getUniversityListByUser(saveUnivStorage, username)

        if (init && univKeyToStore) {
            if (Boolean((saveUniversityByUser as any)[ univKeyToStore ])) {
                setButtonLabel('Remove')
            }

            return
        }

        const newSetOfUnivList = {
            ...univListFromStorageByUser,
            [ username ]: {
                ...saveUniversityByUser,
                [ universityKey ]: (buttonLabel === 'Add') ? university : undefined
            }
        }

        setDataToStorage(saveUnivStorage, newSetOfUnivList)

        return
    }, [ universityKey, buttonLabel, university ])
    const handleGetSpecificUniversity = useCallback(async () => {
        try {
            const { environment } = await import('@config/environment')
            const { data }: APIResponseProps = await axios.get(`/search?name=${ name }`)
            const [ { name: univName, ...rest } ] = data
            const univKeyToStore = univName.replaceAll(' ', environment.delimiter)

            setUniversity({ name: univName, ...rest })
            setUniversityKey(univKeyToStore)
            handleGetListofUniv(true, univKeyToStore)
        } catch (error: any) {
            console.error('Error Found: ', error)
        }
    }, [ name, handleGetListofUniv ])
    const handleOpenWebsite = (website: string) => {
        window.open(website, '_blank')
    }
    const handleToggleUniversityFromStorage = () => {
        handleGetListofUniv(false, '')
        setButtonLabel((prevState: string) => (prevState === 'Add') ? 'Remove' : 'Add')
    }

    useEffect(() => {
        handleGetSpecificUniversity()
    }, [handleGetSpecificUniversity ])

    const buttonStyle = {
        ...styles.button,
        background: `#${ (buttonLabel === 'Add') ? '1a73e8' : 'f44336' }`
    }
    const universityElement = (
        <Fragment>
            <Banner title={ university?.name || '' } />
            <div style={ styles.information }>
                <div style={ styles.save }>
                    <Button style={ buttonStyle } onClick={ handleToggleUniversityFromStorage }>
                        { buttonLabel } University
                    </Button>
                </div>
                <Information
                    title="University"
                    content={ `${ university?.name }, ${ university?.alpha_two_code }` || '' }
                />
                <Information
                    title="Country"
                    content={ university?.country || '' }
                />
                <Information title="Website(s)">
                    { university?.web_pages.map(
                        (website: string) => (
                            <WebsiteLink
                                key={ website }
                                website={ website }
                                onSelect={ handleOpenWebsite }
                            />
                        )
                    ) }
                </Information>
                <Information title="Domain(s)">
                    { university?.domains.map(
                        (domain: string) => (
                            <p key={ domain }>{ domain }</p>
                        )
                    ) }
                </Information>
            </div>
        </Fragment>
    )

    return universityElement
}

export default University