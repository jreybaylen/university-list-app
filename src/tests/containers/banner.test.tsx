import { Suspense } from 'react'
import { MemoryRouter , Route } from 'react-router-dom'
import { render, waitFor, fireEvent } from '@testing-library/react'

import Banner from '@container/Banner'
import University from '@pages/University'

describe('<Banner /> Component', () => {
    it('Should render without crashing', () => {
        const wrapper = render(<Banner title="" />)

        expect(wrapper).toMatchSnapshot()
    })

    it('Should display title content', () => {
        const title = "John Doe"
        const { container, getByText } = render(<Banner title={ title } />)

        expect(getByText(title)).toBeInTheDocument()
        expect(container.querySelector('h1')).toBeInTheDocument()
    })

    it('Should render back button', () => {
        const title = "John Doe"
        const { container, getByText } = render(<Banner title={ title } />)

        expect(getByText(title)).toBeInTheDocument()
        expect(container.querySelector('button')).toBeInTheDocument()
    })

    it('Should trigger redirect to home page when back button has been clicked', async () => {
        const routeCofig = {
            history: {},
            location: {}
        }
        const defaultURL = '/university/Test$@Academy'
        const { getByText } = render(
            <Suspense fallback="">
                <MemoryRouter initialEntries={ [ defaultURL ] }>
                    <Route path="/university/:name" component={ University } />
                    <Route
                        path="*"
                        render={({ history, location }) => {
                            routeCofig.history = history
                            routeCofig.location = location

                            return <div />
                        }}
                    />
                </MemoryRouter>
            </Suspense>
        )

        expect((routeCofig.location as { pathname: string }).pathname).toStrictEqual(defaultURL)
        await waitFor(() => {
            expect(getByText(/back/i)).toBeInTheDocument()
            fireEvent.click(getByText(/back/i))
            expect((routeCofig.location as { pathname: string }).pathname).toStrictEqual('/')
        })
    })
})