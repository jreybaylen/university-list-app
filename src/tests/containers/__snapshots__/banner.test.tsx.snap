// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> Component Should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        style="top: 20px; left: 20px; color: rgb(255, 255, 255); cursor: pointer; border-radius: 25px; padding: 10px 20px; position: absolute; border: 1px solid #fff; background: rgba(0, 0, 0, 0.5);"
      >
        Back
      </button>
      <div
        style="margin-bottom: 50px; padding: 150px 0px; background-size: cover; background-position: center; background-repeat: no-repeat; box-shadow: 0px 0px 15px 5px #ccc; background-image: url(univ-bg.jpg);"
      >
        <h1
          style="width: 80%; color: rgb(255, 255, 255); margin: auto; max-width: 1000px; font-weight: 500; border-radius: 5px; text-align: center; padding: 30px 15px; border: 2px solid #fff; background: rgba(0, 0, 0, 0.6); box-shadow: rgb(0 0 0 / 20%) 0px 4px 8px 0px;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <button
      style="top: 20px; left: 20px; color: rgb(255, 255, 255); cursor: pointer; border-radius: 25px; padding: 10px 20px; position: absolute; border: 1px solid #fff; background: rgba(0, 0, 0, 0.5);"
    >
      Back
    </button>
    <div
      style="margin-bottom: 50px; padding: 150px 0px; background-size: cover; background-position: center; background-repeat: no-repeat; box-shadow: 0px 0px 15px 5px #ccc; background-image: url(univ-bg.jpg);"
    >
      <h1
        style="width: 80%; color: rgb(255, 255, 255); margin: auto; max-width: 1000px; font-weight: 500; border-radius: 5px; text-align: center; padding: 30px 15px; border: 2px solid #fff; background: rgba(0, 0, 0, 0.6); box-shadow: rgb(0 0 0 / 20%) 0px 4px 8px 0px;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
